sympy/solvers/inequalities.py

```diff
@@ -982,3 +987,457 @@ def reduce_inequalities(inequalities, symbols=[]):
+    Matrix([
+    [-3/2, 11/2, 7/2],
+    [ 1/2, -3/2, 1/2],
+    [   0,    3,   1]])
+    """
+    Mi, Mj, Mij = M[i,:], M[:,j], M[i,j]
+    if Mij == 0:
+        raise ZeroDivisionError("Tried to pivot about zero-valued entry.")
+    MM = M - Mj * (Mi / Mij)
+    MM[i,:] = Mi / Mij
+    MM[:,j] = -Mj / Mij
+    MM[i,j] = 1 / Mij
+    return MM
+
+
+def _simplex(M, skip_phase_2=False):
+    """
+    Simplex method with randomized pivoting
+
+    LINEAR PROGRAMMING: A Concise Introduction by Thomas S. Ferguson
+      - http://web.tecnico.ulisboa.pt/mcasquilho/acad/or/ftp/FergusonUCLA_LP.pdf
+
+    Explanation
+    ===========
+
+    When x is a column vector of variables, y is a column vector of dual variables,
+    and when the objective is either:
+
+    *) Maximizing Cx constrained to Ax <= B and x >= 0.
+    *) Minimizing y^{T}B constrained to y^{T}A >= C^{T} and y >= 0.
+
+    The method returns a triplet of solutions optimum, argmax and argmax_dual where
+    optimum is the optimum solution, and argmax/ argmax_dual give the
+    values for x and y, respectively.
+
+    Examples
+    ========
+
+    Suppose we want to maximize
+        3x+y
+    subject to
+        x+y <= 2
+
+    >>> from sympy.matrices.dense import Matrix
+    >>> from sympy.matrices.immutable import ImmutableMatrix
+    >>> from sympy.core.random import seed
+    >>> from sympy.solvers.inequalities import _simplex
+    >>> A = Matrix([[1,1]])
+    >>> B = Matrix([[2]])
+    >>> C = Matrix([[3, 1]])
+    >>> D = ImmutableMatrix([0])
+    >>> M = Matrix([[A, B], [-C, D]])
+    >>> M
+    Matrix([
+    [ 1,  1, 2],
+    [-3, -1, 0]])
+    >>> seed(47)
+    >>> optimum, argmax, argmax_dual = _simplex(M)
+    >>> optimum
+    6
+    >>> argmax
+    [2, 0]
+    >>> argmax_dual
+    [3]
+
+    Suppose we want to maximize
+        3x+y
+    subject to
+        x <= 2
+        y <= 4
+
+    >>> A = Matrix([[1,0], [0, 1]])
+    >>> B = Matrix([[2],[4]])
+    >>> C = Matrix([[3, 1]])
+    >>> D = ImmutableMatrix([0])
+    >>> M = Matrix([[A, B], [-C, D]])
+    >>> M
+    Matrix([
+    [ 1,  0, 2],
+    [ 0,  1, 4],
+    [-3, -1, 0]])
+    >>> optimum, argmax, argmax_dual = _simplex(M)
+    >>> optimum
+    10
+    >>> argmax
+    [2, 4]
+    >>> argmax_dual
+    [3, 1]
+    """
+    rand = random.Random(x=core_random())
+
+    phase = 1
+
+    r_orig = ['x_{}'.format(j) for j in range(M.cols - 1)]
+    s_orig = ['y_{}'.format(i) for i in range(M.rows - 1)]
+
+    R = r_orig.copy()
+    S = s_orig.copy()
+
+    while True:
+        B = M[:-1, -1]
+        A = M[:-1, :-1]
+        C = M[-1, :-1]
+
+        if phase == 1:
+            # In phase 1 we look for a feasible solution. Once we find one, we continue to phase 2.
+            if all(B[i] >= 0 for i in range(B.rows)):
+                if skip_phase_2:
+                    break
+                else:
+                    phase = 2
+                    continue
```

Comment:
This control flow with the phase variable is odd. It would be easier to follow if it was just two loops.

Reply:
If I did it in two loops there would be this duplicate code:

```
        # For each row, we calculate a ratio: B[i] / A[i, j0].
        # The row with the smallest ratio, i0, contains the pivot entry at its intersection with j0.
        # If there are ties, pick a row at random.
        min_ratio = float('inf')
        min_rows = []
        for i in piv_rows:
            ratio = B[i] / A[i, j0]
            if ratio < min_ratio:
                min_ratio = ratio
                min_rows = [i]
            elif ratio == min_ratio:
                min_rows.append(i)
        rand.shuffle(min_rows)
        i0 = min_rows[0]

        M = _pivot(M, i0, j0)
        R[j0], S[i0] = S[i0], R[j0]
```

Reply:
If I were to split it up into two loops, perhaps I could make a helper function for this code to avoid duplicate code.

```
        # For each row, we calculate a ratio: B[i] / A[i, j0].
        # The row with the smallest ratio, i0, contains the pivot entry at its intersection with j0.
        # If there are ties, pick a row at random.
        min_ratio = float('inf')
        min_rows = []
        for i in piv_rows:
            ratio = B[i] / A[i, j0]
            if ratio < min_ratio:
                min_ratio = ratio
                min_rows = [i]
            elif ratio == min_ratio:
                min_rows.append(i)
        rand.shuffle(min_rows)
        i0 = min_rows[0]

        M = _pivot(M, i0, j0)
        R[j0], S[i0] = S[i0], R[j0]
```

Reply:
It is better to have a bit of code duplication than a complicated control flow. If it makes sense to define a helper function and that also makes things more readable then that is often a good way to reduce duplication:
```python
def choose_pivot_row(B, A, rows, col):
    ratios = [B[i] / A[i][col] for i in rows]
    min_ratio = min(ratios)
    min_rows = [i for i in rows if ratios[i] == min_ratio]
    row = rand.choice(min_rows)
    return row
```